snuba_sdk/query.py

```diff
@@ -3,28 +3,33 @@
 from dataclasses import dataclass, replace
 from typing import Any, List, Optional, Sequence, Union
 
-from snuba_sdk.entity import Entity
 from snuba_sdk.expressions import (
     Column,
+    Condition,
+    Entity,
     Function,
     Granularity,
     Limit,
     Offset,
+    Translation,
+    Validation,
 )
-from snuba_sdk.conditions import Condition
 
 
 class InvalidQuery(Exception):
     pass
 
 
 def list_type(vals: List[Any], type_classes: Sequence[Any]) -> bool:
-    print("1", isinstance(vals, list))
     return isinstance(vals, list) and all(
         isinstance(v, tuple(type_classes)) for v in vals
     )
 
 
+TRANSLATOR = Translation()
+VALIDATOR = Validation()
+
+
 @dataclass(frozen=True)
 class Query:
```

Comment:
I cannot remember  if there was a reason for not making Query an expression. Or not having a common class on top of Query and expression (like Visitable) that would allow to use the same approach to navigate through the tree for Queries and expressions. That would mean learning one way instead of two.
In Snuba this was for historic reasons, Query was too tangled with the dataset logic at the time ew introduced the AST. The second reason, that would be present here as well, is that turning a query into an expression is not enough, as we do not want to have an argument of a function call to be a Query, so there has to be an intermediate class.

Reply:
I have a comment in the validation function I think, but the main reason the Query isn't an Expression is validation. An Expression is assumed to be always valid, since I check its validity in the __post_init__. However a Query can exist in an invalid state, if for example you initialize it and then add the select columns later. While waiting for the select columns the query is in an invalid state. Once I add subquery support  the idea of making a Query visitable will be needed.